cpp/diffraction/StaticStructureFactorDirect.cc

```diff
@@ -0,0 +1,306 @@
+// Copyright (c) 2010-2020 The Regents of the University of Michigan
+// This file is from the freud project, released under the BSD 3-Clause License.
+
+#include <algorithm>
+#include <cmath>
+#include <complex>
+#include <limits>
+#include <random>
+#include <stdexcept>
+#include <tbb/concurrent_vector.h>
+
+#include "Eigen/Eigen/Dense"
+
+#include "Box.h"
+#include "ManagedArray.h"
+#include "NeighborQuery.h"
+#include "StaticStructureFactorDirect.h"
+#include "utils.h"
+
+/*! \file StaticStructureFactorDirect.cc
+    \brief Routines for computing static structure factors.
+*/
+
+namespace freud { namespace diffraction {
+
+StaticStructureFactorDirect::StaticStructureFactorDirect(unsigned int bins, float k_max, float k_min,
+                                                         unsigned int max_k_points)
+    : m_max_k_points(max_k_points)
+{
+    if (bins == 0)
+    {
+        throw std::invalid_argument("StaticStructureFactorDirect requires a nonzero number of bins.");
+    }
+    if (k_max <= 0)
+    {
+        throw std::invalid_argument("StaticStructureFactorDirect requires k_max to be positive.");
+    }
+    if (k_min < 0)
+    {
+        throw std::invalid_argument("StaticStructureFactorDirect requires k_min to be non-negative.");
+    }
+    if (k_max <= k_min)
+    {
+        throw std::invalid_argument(
+            "StaticStructureFactorDirect requires that k_max must be greater than k_min.");
+    }
+
+    // We must construct two separate histograms, one for the counts and one
+    // for the actual correlation function. The counts are used to normalize
+    // the correlation function.
+    const auto axes
+        = StructureFactorHistogram::Axes {std::make_shared<util::RegularAxis>(bins, k_min, k_max)};
+    m_structure_factor = StructureFactorHistogram(axes);
+    m_local_structure_factor = StructureFactorHistogram::ThreadLocalHistogram(m_structure_factor);
+    m_histogram = KBinHistogram(axes);
+    m_local_histograms = KBinHistogram::ThreadLocalHistogram(m_histogram);
+}
+
+void StaticStructureFactorDirect::accumulate(const freud::locality::NeighborQuery* neighbor_query,
+                                             const vec3<float>* query_points, unsigned int n_query_points,
+                                             unsigned int n_total)
+{
+    // Compute k vectors by sampling reciprocal space.
+    const auto& box = neighbor_query->getBox();
+    if (box.is2D())
+    {
+        throw std::invalid_argument("2D boxes are not currently supported.");
+    }
+    const auto k_bin_edges = m_structure_factor.getBinEdges()[0];
+    const auto k_min = k_bin_edges.front();
+    const auto k_max = k_bin_edges.back();
+    if ((!box_assigned)
+        || (box != previous_box && box.getPeriodicX() != previous_box.getPeriodicX()
```

Comment:
Refactor the equality operator instead: https://github.com/glotzerlab/freud/blob/9b4db2bf4a073c0c3483de6cccafcecadc0c8efc/cpp/box/Box.h#L82